### PR TITLE
Rsa review

### PR DESCRIPTION
**Pull request by CharNoir**
- The code is well-structured.
- The code is mostly self-explanatory, but adding comments to complex logic sections would improve readability and maintainability.
- A lot of code parts are inefficient.

### CODE REVIEW

File: algorithms/maths/rsa.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# n,e,d = generate_key(16)
+# data = 20
+# encrypted = pow(data,e,n)
+# decrypted = pow(encrypted,d,n)
+# assert decrypted == data
+
+import random
+
+
+def generate_key(k, seed=None):
```
Comment (CharNoir) [inline algorithms/maths/rsa.py:34]: Consider using a more efficient primality test like Miller-Rabin for generating large primes, especially for longer key sizes.

File: algorithms/maths/rsa.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    the RSA key generating algorithm
+    k is the number of bits in n
+    """
+
+    def modinv(a, m):
```
Comment (CharNoir) [inline algorithms/maths/rsa.py:40]: Use an efficient algorithm like the Extended Euclidean Algorithm to compute modular inverses.

Comment (CharNoir) [reply algorithms/maths/rsa.py:40]: Example of possible code:

`def extended_gcd(a, b):
    if b == 0:
        return a, 1, 0
    gcd, x1, y1 = extended_gcd(b, a % b)
    x = y1
    y = x1 - (a // b) * y1
    return gcd, x, y

def mod_inverse(a, m):
    gcd, x, y = extended_gcd(a, m)
    if gcd != 1:
        raise ValueError(f"The modular inverse of {a} modulo {m} does not exist.")
    else:
        return x % m`
        
This code is not perfect due to using recursion, however it allows to boost efficiency.

File: algorithms/maths/rsa.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        """generate a prime with k bits"""
+
+        def is_prime(num):
+            if num == 2:
+                return True
+            for i in range(2, int(num ** 0.5) + 1):
+                if num % i == 0:
+                    return False
+            return True
+
```
Comment (CharNoir) [inline algorithms/maths/rsa.py:58]: Consider using a more efficient primality test like Miller-Rabin for generating large primes, especially for longer key sizes.

File: algorithms/maths/rsa.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        random.seed(seed)
+        while True:
+            key = random.randrange(int(2 ** (k - 1)), int(2 ** k))
```
Comment (CharNoir) [inline algorithms/maths/rsa.py:61]: Consider using secrets module for cryptographic purposes instead of random for generating secure random numbers.

**Review by CharNoir — COMMENTED** [review]
For serious cryptography, enhance RSA algorithm with recommended improvements  before deploying them in production or critical applications.

**Review by retroruiner — APPROVED** [review]

**Review by retroruiner — APPROVED** [review]

**Pull request merged**